superbench/benchmarks/micro_benchmarks/computation_communication_overlap.py

```diff
@@ -0,0 +1,242 @@
+            matB (tensor): the matrix used in matmul
+            stages (int): the ratio number of computation kernel and communication kernel
+            message(tensor): the data used to be transferred through NCCL
+            times(int): number of times in one step to run
+
+        Return:
+            True of False: if computation kernel type is invalid, return False, else, return True
+        """
+        for i in range(times):
+            if self.__world_size > 1:
+                torch.distributed.all_reduce(message, op=torch.distributed.ReduceOp.SUM, async_op=True)
+            for stage in range(stages):
+                if kernel == ComputationKernelType.MUL:
+                    matA[stage].mul(matA[stage])
+                elif kernel == ComputationKernelType.MATMUL:
+                    matA[stage].matmul(matB)
+                else:
+                    logger.error(
+                        'Unknown comoputation kernel type - benchmark: {}, type: {}.'.format(self._name, kernel)
+                    )
+                    return False
+        return True
+
+    def _benchmark(self):
+        """Implementation for benchmarking."""
+        M = self._args.m
+        K = self._args.k
+        N = self._args.n
+        P = self._args.p
+        Q = self._args.q
+        kernel = self._args.kernel
+        if self.__local_rank == 0:
+            logger.info(
+                'Computation Communication Overlap - using {} GPUs,\
+                matrix shape for computation: M={} K={} N={},\
+                message tensor shape of nccl = [{},{}],\
+                ratio between computation kernel and nccl kernel={}'.format(
+                    self.__world_size, M, K, N, P, Q, self._args.ratio
+                )
+            )
+
+        MatA = list()
+        MatB = list()
+        # Matrix A
+        for _ in range(self._args.ratio):
+            MatA.append(torch.randn(M, K).cuda())
+        # Matrix B
+        MatB = torch.randn(K, N).cuda()
+        # message for nccl to transport
+        shape = [P, Q]
+        message = torch.randn(*shape).cuda()
+
+        # warm up
+        for i in range(self._args.num_warmup):
+            if not self.__kernel_nccl_pipeline(kernel, MatA, MatB, self._args.ratio, message, times=100):
+                return False
+        torch.cuda.synchronize()
+
+        # run and collect results
+        start = time.perf_counter()
+        for i in range(self._args.num_steps):
+            self.__kernel_nccl_pipeline(kernel, MatA, MatB, self._args.ratio, message, times=100)
+        compute_end = time.perf_counter()
+        torch.cuda.synchronize()
+        end = time.perf_counter()
+
+        logger.info(
+            'GPU {} total compute cost {}'.format(
+                self.__local_rank, (compute_end - start) * 1000 / self._args.num_steps
+            )
+        )
+
+        compute_metric = 'computation_communication_overlap_gpu{}_computation_cost_{}'.format(self.__local_rank, kernel)
```

Comment:
Why do we need two similiar metrics? Which metric will show the unusual results? or both can?

Reply:
compute_metric is the metric for computation end, if any gpu has issue, it will show unusual for different gpu. Another metric is for both computation and communication end, it is almost same for all gpus in a machine at any time, it's a more stable value, so in my opinion it's used to check if this machine has performance downgrade for different machines.

Reply:
Thanks! Let's have a discussion on monday.